File: leetcode/404.sum-of-left-leaves.ts
/*
 * @lc app=leetcode id=404 lang=typescript
 *
 * [404] Sum of Left Leaves
 */

class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}
// @lc code=start
function sumOfLeftLeaves(root: TreeNode | null): number {
  let ans = 0;

  const dfs = (node: TreeNode | null, isLeft: boolean): boolean => {
    if (node === null) return true;

    const left = dfs(node.left, true);
    const right = dfs(node.right, false);

    if (left && right && isLeft) ans += node.val;
  }
  dfs(root, false);

  return ans;
};
// @lc code=end

File: leetcode/404.sum-of-left-leaves.cpp
/*
 * @lc app=leetcode id=404 lang=cpp
 *
 * [404] Sum of Left Leaves
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int dsf(TreeNode *node, bool isLeft)
    {
        if (node == nullptr)
            return 0;

        if (node->left == nullptr && node->right == nullptr && isLeft)
        {
            return node->val;
        }
        int l = dsf(node->left, true);
        int r = dsf(node->right, false);

        return l + r;
    }
    int sumOfLeftLeaves(TreeNode *root)
    {
        return dsf(root, false);
    }
};
// @lc code=end
